src/ucp/core/ucp_context.c

```diff
@@ -666,6 +669,31 @@ static ucs_status_t ucp_fill_resources(ucp_context_h context,
          * client-server connection establishment via sockaddr, don't use it */
         if ((num_tl_resources > 0) ||
             (context->tl_mds[md_index].attr.cap.flags & UCT_MD_FLAG_SOCKADDR)) {
+            /*List of memory type MDs*/
+            if (context->tl_mds[md_index].attr.cap.mem_type != UCT_MD_MEM_TYPE_HOST) {
+                if (context->mem_type_tl_mds == NULL) {
+                    context->mem_type_tl_mds = ucs_malloc(sizeof(*context->mem_type_tl_mds) *
+                                                          UCT_MD_MEM_TYPE_LAST,
+                                                          "ucp_mem_type_tl_mds");
+                    if (context->mem_type_tl_mds == NULL) {
+                        status = UCS_ERR_NO_MEMORY;
+                        goto err_free_context_resources;
+                    }
+                }
+                /* check if mem type MD is already in the list */
+                j = 0;
+                while (j < context->num_mem_type_mds) {
```

Comment:
maybe have a temporary bitmap of mem_types which are already "covered" to not go over the array again? performance is not that important here, but it will also simplify the code

Reply:
Fixed